File: frontend/src/utils/attendanceNotes.ts
/**
 * Chuẩn hoá ghi chú ngày công (chuỗi nối bằng `;`) để hiển thị rõ ràng.
 */

export type AttendanceNoteKind =
  | 'sync'
  | 'deployment_ot'
  | 'deployment_inside'
  | 'leave'
  | 'unpaid'
  | 'business_trip'
  | 'explanation'
  | 'update'
  | 'other';

export type ParsedAttendanceNote = {
  kind: AttendanceNoteKind;
  /** Tiêu đề ngắn (vd. Điều động làm thêm ×1.5) */
  title: string;
  /** Khung giờ (vd. 06:20–06:45) */
  timeRange?: string;
  /** Phần giờ thực → giờ công (vd. 0.42h → 0.63h công) */
  hoursLine?: string;
  /** Phân bổ sáng/chiều/ngoài giờ */
  breakdown?: string;
  /** Lý do / mô tả công việc */
  reason?: string;
  /** Mã tham chiếu đơn (vd. DD:54) */
  ref?: string;
  /** true nếu là điều động đã gắn mã đơn duyệt */
  approved?: boolean;
  raw: string;
};

const REF_RE = /\[(DD(?:TC)?:\d+)\]\s*$/i;
const TIME_RANGE_RE = /(\d{1,2}:\d{2}\s*[–\-—]\s*\d{1,2}:\d{2})/;
const BREAKDOWN_RE = /\(([^)]*sáng[^)]*)\)/i;
const COEFF_RE = /×\s*([\d.,]+)/;

function classifySimple(text: string): AttendanceNoteKind | null {
  const t = text.toLowerCase();
  if (t.includes('đồng bộ máy chấm')) return 'sync';
  if (t.includes('nghỉ phép') || t.startsWith('phép')) return 'leave';
  if (t.includes('không lương')) return 'unpaid';
  if (t.includes('công tác')) return 'business_trip';
  if (t.includes('giải trình')) return 'explanation';
  if (t.includes('cập nhật công')) return 'update';
  return null;
}

function stripRef(text: string): { body: string; ref?: string } {
  const m = text.match(REF_RE);
  if (!m || m.index == null) return { body: text.trim() };
  return {
    body: text.slice(0, m.index).trim(),
    ref: m[1],
  };
}

function parseDeployment(body: string, kind: 'deployment_ot' | 'deployment_inside', defaultTitle: string): ParsedAttendanceNote {
  const coeff = body.match(COEFF_RE)?.[1];
  const title = coeff ? `${defaultTitle} ×${coeff}` : defaultTitle;
  const timeRange = body.match(TIME_RANGE_RE)?.[1]?.replace(/\s+/g, ' ').trim();

  let breakdown: string | undefined;
  const bd = body.match(BREAKDOWN_RE);
  if (bd) breakdown = bd[1].trim();

  let hoursLine: string | undefined;
  if (timeRange) {
    const afterTime = body.slice(body.indexOf(timeRange) + timeRange.length);
    const hoursMatch = afterTime.match(/·\s*([^(]+?)(?=\s*\(|$)/);
    if (hoursMatch) {
      hoursLine = hoursMatch[1].trim().replace(/\s*[—-]\s*$/, '').trim();
      if (!hoursLine) hoursLine = undefined;
    } else {
      // Điều động trong ca: "... — 0.5 công (...)"
      const dashHours = afterTime.match(/[—-]\s*([^(]+?)(?=\s*\(|$)/);
      if (dashHours) {
        hoursLine = dashHours[1].trim();
        if (!hoursLine) hoursLine = undefined;
      }
    }
  }

  let reason: string | undefined;
  if (breakdown) {
    const afterBd = body.slice(body.indexOf(`(${breakdown})`) + breakdown.length + 2).trim();
    reason = afterBd.replace(/^:\s*/, '').trim() || undefined;
  } else {
    const colonParts = body.split(':');
    if (colonParts.length >= 3) {
      // title: time... : reason  OR title: rest
      const last = colonParts[colonParts.length - 1].trim();
      // Avoid treating time fragments as reason
      if (last && !/^\d{1,2}$/.test(last) && !TIME_RANGE_RE.test(last)) {
        // If last segment looks like a reason (has letters beyond time/hours tokens)
        if (/[A-Za-zÀ-ỹ]/.test(last) && !/^\d/.test(last.replace(/^[·\s]+/, ''))) {
          reason = last;
        }
      }
    }
  }

  return {
    kind,
    title,
    timeRange,
    hoursLine,
    breakdown,
    reason,
    raw: body,
  };
}

function parseOne(rawPart: string): ParsedAttendanceNote | null {
  const raw = rawPart.trim();
  if (!raw) return null;

  const { body, ref } = stripRef(raw);

  if (/^điều động làm thêm/i.test(body)) {
    return {
      ...parseDeployment(body, 'deployment_ot', 'Điều động làm thêm'),
      ref,
      approved: Boolean(ref) || /\[DD(?:TC)?:/i.test(body),
      raw,
    };
  }

  if (/^điều động trong ca/i.test(body)) {
    return {
      ...parseDeployment(body, 'deployment_inside', 'Điều động trong ca'),
      ref,
      approved: Boolean(ref) || /\[DD(?:TC)?:/i.test(body),
      raw,
    };
  }

  const simple = classifySimple(body);
  if (simple === 'sync') {
    return { kind: 'sync', title: 'Đồng bộ máy chấm công', ref, raw };
  }
  if (simple) {
    return { kind: simple, title: body, ref, raw };
  }

  return { kind: 'other', title: body, ref, raw };
}

/** Tách ghi chú ngày công thành các mục đọc được. */
export function parseAttendanceNotes(note?: string | null): ParsedAttendanceNote[] {
  if (!note?.trim()) return [];
  return note
    .split(';')
    .map((p) => parseOne(p))
    .filter((x): x is ParsedAttendanceNote => Boolean(x));
}

export function attendanceNoteKindLabel(kind: AttendanceNoteKind): string {
  switch (kind) {
    case 'sync':
      return 'Đồng bộ';
    case 'deployment_ot':
      return 'Điều động OT';
    case 'deployment_inside':
      return 'Điều động trong ca';
    case 'leave':
      return 'Nghỉ phép';
    case 'unpaid':
      return 'Không lương';
    case 'business_trip':
      return 'Công tác';
    case 'explanation':
      return 'Giải trình';
    case 'update':
      return 'Cập nhật công';
    default:
      return 'Ghi chú';
  }
}

/** Phiên bản text gọn (tooltip / bảng tháng). */
export function formatAttendanceNotesPlain(note?: string | null): string {
  const items = parseAttendanceNotes(note);
  if (!items.length) return note?.trim() || '';
  return items
    .map((n) => {
      const bits = [n.title];
      if (n.timeRange) bits.push(n.timeRange);
      if (n.hoursLine) bits.push(n.hoursLine);
      if (n.reason) bits.push(n.reason);
      if (n.ref) bits.push(`[${n.ref}]`);
      return bits.join(' · ');
    })
    .join('\n');
}

File: mobile/lib/core/utils/attendance_notes.dart
/// Parse ghi chú ngày công (nối bằng `;`) để hiển thị gọn trên app.
class ParsedAttendanceNote {
  const ParsedAttendanceNote({
    required this.kind,
    required this.title,
    this.timeRange,
    this.hoursLine,
    this.breakdown,
    this.reason,
    this.ref,
    required this.raw,
  });

  final String kind;
  final String title;
  final String? timeRange;
  final String? hoursLine;
  final String? breakdown;
  final String? reason;
  final String? ref;
  final String raw;
}

final _refRe = RegExp(r'\[(DD(?:TC)?:\d+)\]\s*$', caseSensitive: false);
final _timeRe = RegExp(r'(\d{1,2}:\d{2}\s*[–\-—]\s*\d{1,2}:\d{2})');
final _breakdownRe = RegExp(r'\(([^)]*sáng[^)]*)\)', caseSensitive: false);
final _coeffRe = RegExp(r'×\s*([\d.,]+)');

List<ParsedAttendanceNote> parseAttendanceNotes(String? note) {
  final raw = note?.trim();
  if (raw == null || raw.isEmpty) return const [];
  return raw
      .split(';')
      .map((p) => _parseOne(p.trim()))
      .whereType<ParsedAttendanceNote>()
      .toList();
}

ParsedAttendanceNote? _parseOne(String raw) {
  if (raw.isEmpty) return null;
  final refMatch = _refRe.firstMatch(raw);
  final ref = refMatch?.group(1);
  final body = refMatch == null
      ? raw
      : raw.substring(0, refMatch.start).trim();
  final lower = body.toLowerCase();

  if (lower.startsWith('điều động làm thêm')) {
    return _parseDeployment(body, raw, ref, 'deployment_ot', 'Điều động làm thêm');
  }
  if (lower.startsWith('điều động trong ca')) {
    return _parseDeployment(body, raw, ref, 'deployment_inside', 'Điều động trong ca');
  }
  if (lower.contains('đồng bộ máy chấm')) {
    return ParsedAttendanceNote(
      kind: 'sync',
      title: 'Đồng bộ máy chấm công',
      ref: ref,
      raw: raw,
    );
  }
  return ParsedAttendanceNote(kind: 'other', title: body, ref: ref, raw: raw);
}

ParsedAttendanceNote _parseDeployment(
  String body,
  String raw,
  String? ref,
  String kind,
  String defaultTitle,
) {
  final coeff = _coeffRe.firstMatch(body)?.group(1);
  final title = coeff == null ? defaultTitle : '$defaultTitle ×$coeff';
  final timeRange = _timeRe.firstMatch(body)?.group(1)?.replaceAll(RegExp(r'\s+'), ' ');
  final breakdown = _breakdownRe.firstMatch(body)?.group(1)?.trim();

  String? hoursLine;
  if (timeRange != null) {
    final idx = body.indexOf(timeRange);
    final after = body.substring(idx + timeRange.length);
    final dot = RegExp(r'·\s*([^(]+?)(?=\s*\(|$)').firstMatch(after);
    if (dot != null) {
      hoursLine = dot.group(1)?.trim();
    } else {
      final dash = RegExp(r'[—-]\s*([^(]+?)(?=\s*\(|$)').firstMatch(after);
      hoursLine = dash?.group(1)?.trim();
    }
    if (hoursLine != null && hoursLine.isEmpty) hoursLine = null;
  }

  String? reason;
  if (breakdown != null) {
    final token = '($breakdown)';
    final idx = body.indexOf(token);
    if (idx >= 0) {
      reason = body.substring(idx + token.length).trim().replaceFirst(RegExp(r'^:\s*'), '');
      if (reason.isEmpty) reason = null;
    }
  }

  return ParsedAttendanceNote(
    kind: kind,
    title: title,
    timeRange: timeRange,
    hoursLine: hoursLine,
    breakdown: breakdown,
    reason: reason,
    ref: ref,
    raw: raw,
  );
}
